Windows loadlibrary

guschmue opened the pull request:
windows:
- support for tf.load_library()
- support for linking against tensorflow.dll to access c/c++ api
- switched contrib/rnn to use .so's
- enabled all tests in contrib/rnn

Commit:
cmake versions > 3.6 will quote -D__VERSION__=\"MSVC\" as -D"__VERSION__=\"MSVC\""
which nvcc stubles over. Since this is only used by version.cc we apply this only
core/framework so nvcc will never see it.

Commit:
- support for tf.load_library() on windows
- support for linking against tensorflow.dll to access c/c++ api
- switched contrib/rnn to use .so's
- added add tests in contrib/rnn

tensorflow-jenkins commented:
Can one of the admins verify this patch?

guschmue commented:
this should enable https://github.com/tensorflow/models/issues/1103

tomwanzek commented:
@guschmue am I assuming that this PR might also address the gap seemingly underlying #6184, i.e. see specifically [this issue comment](https://github.com/tensorflow/tensorflow/issues/6184#issuecomment-267374882)

Or am I overinterpreting here? Thx to all for the hard work on TensorFlow!

guschmue commented:
For now I added only the .so's in contrib/rnn to keep this pr reasonable small (fyi @mrry).
But now it should be real easy to add a .so for windows  and we can make a pass over what should be included.

mrry commented on tensorflow/contrib/cmake/tf_python.cmake:
```diff
@@ -597,6 +592,55 @@ add_library(pywrap_tensorflow_internal SHARED
+        $<$<BOOL:${tensorflow_ENABLE_GRPC_SUPPORT}>:$<TARGET_OBJECTS:tf_core_distributed_runtime>>
+        $<TARGET_OBJECTS:tf_core_kernels>
+        $<$<BOOL:${tensorflow_ENABLE_GPU}>:$<TARGET_OBJECTS:tf_core_kernels_cpu_only>>
+        $<$<BOOL:${tensorflow_ENABLE_GPU}>:$<TARGET_OBJECTS:tf_stream_executor>>
```
Suggestion: since we use this list again in the definition of `pywrap_tensorflow_internal`, should (can?) we extract this list of dependencies into a separate variable and use it in both places?

guschmue replied:
yes, can do

guschmue replied:
Sorry, TARGET_OBJECTS doesn't let me do it.

mrry replied:
Thanks for trying!

mrry commented on tensorflow/contrib/cmake/tf_python.cmake:
```diff
@@ -597,6 +592,55 @@ add_library(pywrap_tensorflow_internal SHARED
+    # be exported from tensorflow.dll. Because there is a limit of 64K sybmols
+    # that can be exported, we filter the symbols with a python script to the namespaces
+    # we need.
+    #
```
Just so I've got the process correct in my head: we're building a static `.lib` in order to get something that we can feed to `dumpbin` and then a custom Python script, ultimately to get a `.def` file. Then we throw away the `.lib` and pass the `.def` to a second `link` command, along with all the same objects, and we get a `.dll` with the appropriate exports?

guschmue replied:
yes. I kept the static lib around - might be handy for building standalone apps.
I use dumpbin /linkermemer to get the symbols - decorated sysbols, hard to read. I run them though undname.exe to get the undecorated names so one can use a readable regex

mrry commented on tensorflow/contrib/cmake/tools/create_deffile.py:
```diff
@@ -0,0 +1,107 @@
+        dupes = 0
+        proc = Popen([UNDNAME, tmpfile.name], stdout=PIPE)
+        for idx, line in enumerate(io.TextIOWrapper(proc.stdout, encoding="utf-8")):
+            decorared = candidates[idx]
```
s/decorared/decorated/?

guschmue replied:
ok

mrry commented on tensorflow/contrib/cmake/tools/create_deffile.py:
```diff
@@ -0,0 +1,107 @@
+INCLUDE_RE = re.compile(r"^(TF_\w*)$|" +
+                        r"tensorflow::|" +
+                        r"functor::|" +
+                        r"perftools::gputools")
```
@keveman: Can you take a look at this list of excluded/included symbols and check that it fits with your idea of what needs to be exported?

mrry commented on tensorflow/contrib/cmake/tf_cc_ops.cmake:
```diff
@@ -120,3 +120,41 @@ list(REMOVE_ITEM tf_cc_srcs ${tf_cc_test_srcs})
+    # on windows copy .dll to .so so the python code does not need to deal with it
+    add_custom_command(TARGET ${_AT_TARGET} POST_BUILD
+        COMMAND ${CMAKE_COMMAND} -E copy $<TARGET_FILE:${_AT_TARGET}>
+            ${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_BUILD_TYPE}/${_AT_TARGET}.so)
```
Minor stylistic nit: should we generate files with a `.dll` extension on Windows, and make `resource_loader` handle that?

guschmue replied:
I have been back and forth on this. The python libs pass in  something.so and I did not want to do a s/.so/.dll/ but can certainly change it.

mrry replied:
We can certainly defer this decision. There might be backwards compatibility consequences for changing how the `resource_loader` functions work....

guschmue replied:
Let me change it to .dll and map it in the resource_loader.

mrry commented on tensorflow/contrib/cmake/tf_cc_ops.cmake:
```diff
@@ -120,3 +120,41 @@ list(REMOVE_ITEM tf_cc_srcs ${tf_cc_test_srcs})
 add_library(tf_cc OBJECT ${tf_cc_srcs})
 add_dependencies(tf_cc tf_cc_framework tf_cc_ops)
+
+set (pywrap_tensorflow_lib "${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_BUILD_TYPE}/pywrap_tensorflow_internal.lib")
```
Checking my understanding: is this the import `.lib` for `pywrap_tensorflow_internal.dll`? If so, do we need to include it in the PIP package so that third parties can write extension modules? (And can you add a comment here to explain this as well?)

guschmue replied:
yes, it should be in the pip but it adds around 35MB (uncompressed) so I wasn't sure.

mrry replied:
In that case, let's leave it out for now, but add a `TODO` comment mentioning that we should add it later to support third-party extensions.

guschmue replied:
compresses down to 1.5MB. 
I think I add the comment for now and we make a separate 'sdk' pass to make it easy to use.

mrry reviewed (changes requested):
Thanks for sending this over, Guenther! This is the biggest missing feature on Windows, and it'll be great to start supporting third-party extensions :).

I have a few clarifying questions/comments about the approach, but nothing major.

/cc FYI: @keveman @asimshankar @drpngx @meteorcloudy

drpngx commented on tensorflow/contrib/cmake/tools/create_deffile.py:
```diff
@@ -0,0 +1,107 @@
+    # Run the symbols through undname to get their undecorated name
+    # so we can filter on something readable.
+    with open(args.output, "w") as def_fp:
+        # track dupes
```
Capitalize comment, here and below. The style guide that we are using mandates 2 spaces indent, instead of 4.

guschmue replied:
fix, used your pylintrc

drpngx commented on tensorflow/core/framework/types.h:
```diff
@@ -68,9 +68,17 @@ class DeviceType {
-extern const char* const DEVICE_CPU;   // "CPU"
-extern const char* const DEVICE_GPU;   // "GPU"
-extern const char* const DEVICE_SYCL;  // "SYCL"
+#if !defined(TF_EXTERN)
```
Maybe this is a better place: https://www.github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/macros.h#L42

Perhaps `TF_DLLEXPORT` is a more standard name:
```
#ifdef TF_COMPILE_LIBRARY
# define TF_DLLEXPORT __declspec(dllimport)
#else
# define TF_DLLEXPORT __declspec(dllexport)
#endif   // TF_COMPILE_LIBRARY
```

In time, we'll export the symbols (whitelist) rather than blacklist.

CC: @skye

guschmue replied:
Like it. Want to call it TF_DLLEXPORT or TF_EXPORT ? Later might be nicer if you want to control visibility on linux.

drpngx commented on tensorflow/contrib/cmake/tools/create_deffile.py:
```diff
@@ -0,0 +1,107 @@
+"""
+create-datafiles.py - tool to create a windows def file to export
+  symbols from tensorflow.dll to enable tf.load_library().
+  Because the linker allows only 64K sumbols to be exported per dll
```
symbols

guschmue replied:
fixed

drpngx reviewed (changes requested):
Really nice.

guschmue commented on tensorflow/contrib/cmake/tools/create_deffile.py:
```diff
@@ -0,0 +1,107 @@
+# external tools we use that come with visual studio sdk
+# we assume that the caller has the correct PATH to the sdk
+UNDNAME = "undname.exe"
+DUMPBIN = "dumpbin.exe"
```
I need both ... mangled to write into the .def and demangled to compare against.

guschmue commented on tensorflow/contrib/cmake/tools/create_deffile.py:
```diff
@@ -0,0 +1,107 @@
+import tempfile
+from subprocess import Popen, PIPE
+
+# external tools we use that come with visual studio sdk
```
cmake needs to be called from a visual studio console window so the path is setup correctly to find dumpbin and undname.

guschmue commented on tensorflow/contrib/cmake/tools/create_deffile.py:
```diff
@@ -0,0 +1,107 @@
+"""
+create-datafiles.py - tool to create a windows def file to export
```
done

guschmue commented on tensorflow/contrib/cmake/tf_core_kernels.cmake:
```diff
@@ -36,11 +36,11 @@ if(tensorflow_BUILD_CONTRIB_KERNELS)
+      #"${tensorflow_source_dir}/tensorflow/contrib/rnn/kernels/gru_ops.cc"
+      #"${tensorflow_source_dir}/tensorflow/contrib/rnn/kernels/lstm_ops.cc"
+      #"${tensorflow_source_dir}/tensorflow/contrib/rnn/ops/gru_ops.cc"
+      #"${tensorflow_source_dir}/tensorflow/contrib/rnn/ops/lstm_ops.cc"
```
my bad. I exclude them now by adding them to tf_core_kernels_windows_exclude_srcs

drpngx commented:
TF_EXPORT sounds good!

On Mar 8, 2017 4:40 PM, "guschmue" <notifications@github.com> wrote:

> *@guschmue* commented on this pull request.
> ------------------------------
>
> In tensorflow/core/framework/types.h
> <https://github.com/tensorflow/tensorflow/pull/8217#discussion_r105061992>
> :
>
> > @@ -68,9 +68,17 @@ class DeviceType {
>  std::ostream& operator<<(std::ostream& os, const DeviceType& d);
>
>  // Convenient constants that can be passed to a DeviceType constructor
> -extern const char* const DEVICE_CPU;   // "CPU"
> -extern const char* const DEVICE_GPU;   // "GPU"
> -extern const char* const DEVICE_SYCL;  // "SYCL"
> +#if !defined(TF_EXTERN)
>
> Like it. Want to call it TF_DLLEXPORT or TF_EXPORT ? Later might be nicer
> if you want to control visibility on linux.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/8217#discussion_r105061992>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AT_SbcUFjNUYjirlXtZ97Q_-jLhrPdsgks5rj0pygaJpZM4MXVGg>
> .
>

Commit:
Address review feedback. Changes:
- Some style changes
- TF_COMPILE_LIBRARY define controls exports from the tensorflow.dll.
  If TF_COMPILE_LIBRARY is defined we compile tensorflow, if undefined
  we compile a user op.
- On windows user ops have now a .dll extension. The resource loader will
  take care of finding it.
- pywrap_tensorflow_internal.lib that is needed to build a user op on windows
  comes with the wheel

drpngx commented:
Jenkins, test this please.

mrry reviewed (approved):
This looks great to me, thanks!

mrry commented:
@tensorflow-jenkins test this please.

Commit:
fix linux build

guschmue commented:
soory, just saw that there was a ')' missing for the Linux path

drpngx commented:
Jenkins, test this please.

Commit:
fix linux build

mrry commented:
@tensorflow-jenkins test this please.

mrry commented:
It looks like the `check_futures_test` is unhappy because the new Python script doesn't have the appropriate copyright and `from future import ...` lines at the top of the file. Can you add the following copyright comment above the module comment:

```python
# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
```

...and the following imports after the module comment:

```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
```

Commit:
keep //tensorflow/tools/test:check_futures_test happy

drpngx commented:
Jenkins, test this please.

Commit:
add copyright

guschmue commented:
just added copyright and futures

drpngx commented:
Jenkins, test this please.

drpngx commented:
Yay!

guschmue commented:
Super. As always you guys are a pleasure to work with.

techscientist commented:
Hi, I'm currently using Tensorflow version 1.1 on Windows 10, but I'm still having this problem: that is, the libraries are not being loaded in Windows...This causes an error when trying to use crossed_columns or bucketize_columns for wide and deep learning. Can you please help me with this?

guschmue commented:
This fix is the master but not in the release branch yet (there was no release since the fix).
You could build from source or fetch a master build from ci.

tomwanzek commented:
@techscientist and @guschmue I just pulled down the last stable nightly build for Windows/GPU (114) (the later builds up to 122 have failed for GPU) and the Windows/CPU build (122). I can confirm that, the original issue underlying #6184 still persists.

By implication `feature_column`, `crossed_column` etc. still fail on Windows. I will update the referenced issue with  a new comment. Would be great, if this could be addressed in a follow-up to this PR as it seems fundamental.

guschmue commented:
ok, I can take a look at this.

tomwanzek commented:
@guschmue Thx a lot. FYI I pinged @mrry on my recent update to #6184 , as I saw his handle mentioned in the diagnosis and a TODO in the code.
